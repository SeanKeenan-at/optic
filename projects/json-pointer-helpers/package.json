{
  "name": "@useoptic/json-pointer-helpers",
  "version": "0.21.6",
  "packageManager": "yarn@3.0.2",
  "main": "build/index.js",
  "types": "build/index.d.ts",
  "files": [
    "/build"
  ],
  "scripts": {
    "dev:test": "jest --colors",
    "build": "yarn tsc --build --verbose",
    "ws:build": "yarn build",
    "ws:test": "yarn dev:test",
    "ws:clean": "rm -rf build"
  },
  "devDependencies": {
    "babel-jest": "26.6.0",
    "jest": "^26.6.0",
    "prettier": "^2.4.1",
    "ts-node": "^10.3.1",
    "typescript": "^4.4.4"
  },
  "jest": {
    "testPathIgnorePatterns": [
      "build"
    ]
  },
  "dependencies": {
    "jsonpointer": "^5.0.0"
  }
}
