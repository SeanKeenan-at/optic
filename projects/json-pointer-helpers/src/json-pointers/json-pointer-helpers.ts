import * as jsonPointer from 'json-pointer';

function append(pointer: string, ...property: string[]): string {
  const parsed = jsonPointer.parse(pointer.toString()) || [];
  return jsonPointer.compile([...parsed, ...property]);
}

function pop(pointer: string): string {
  const parsed = jsonPointer.parse(pointer.toString()) || [];
  parsed.pop();
  return jsonPointer.compile([...parsed]);
}

function unescapeUriSafePointer(inputFromApiToolkit: string): string {
  return decodeURIComponent(inputFromApiToolkit);
}

function compile(input: string[]) {
  return jsonPointer.compile(input);
}

function decode(pointer: string): string[] {
  return jsonPointer.parse(pointer);
}

function tryGet(
  input: any,
  pointer: string
): { match: true; value: any } | { match: false; error: string } {
  try {
    const value = jsonPointer.get(input, pointer);
    return { match: true, value };
  } catch (e: any) {
    return { match: false, error: e.message };
  }
}

export default {
  append,
  pop,
  decode,
  unescapeUriSafePointer,
  get: jsonPointer.get,
  tryGet,
  compile,
};
