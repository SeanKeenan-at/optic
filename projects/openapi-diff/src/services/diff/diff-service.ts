import { IDiff, IDiffService } from './types';
import { OpenAPIDiffingQuestions } from '../read/types';
import { ApiTraffic } from '../traffic/types';
import { shouldDiffAgainstThisSpec } from './differs/should-diff';
import { urlPathDiff, urlPathDiffFromSpec } from './differs/url-path-diff';
import { JsonSchemaJsonDiffer } from './differs/json-schema-json-diff/types';
import { opticJsonSchemaDiffer } from './differs/json-schema-json-diff';
import { pathMethodOperationDiffer } from './differs/path-method-operation-diff';
import { responsesDiffer } from './differs/responses';

export function createDiffServiceWithCachingProjections(
  spec: OpenAPIDiffingQuestions,
  jsonSchemaDiffer: JsonSchemaJsonDiffer = opticJsonSchemaDiffer()
): IDiffService {
  const shouldDiff = shouldDiffAgainstThisSpec(spec);
  const pathMatcher = urlPathDiffFromSpec(spec);
  const operationMatcher = pathMethodOperationDiffer(spec);
  const responsesMatcher = responsesDiffer(spec, jsonSchemaDiffer);

  return {
    jsonSchemaDiffer,
    compare: async (
      traffic: ApiTraffic
    ): Promise<{ diffs: IDiff[]; errors: string[] }> => {
      /*
        This is nested, feels ugly, but probably is the best way to represent a nested data structure that can be diffed...for now
        open to suggestions
       */
      const continueWithDiff =
        shouldDiff.responseDiffsForTraffic(traffic).isMatch;
      if (!continueWithDiff) return { diffs: [], errors: [] };

      const matchesPath = pathMatcher.compareToPath(
        traffic.method,
        traffic.path
      );

      // here we know we've matched an operation
      if (matchesPath.isMatch) {
        const matchesOperation =
          operationMatcher.comparePathAndMethodToOperation(
            matchesPath.context,
            traffic.method
          );

        if (matchesOperation.isMatch) {
          const results: { diffs: IDiff[]; errors: string[] } = {
            diffs: [],
            errors: [],
          };
          // match the response
          const response = responsesMatcher.responseDiffsForTraffic(
            traffic,
            matchesOperation.context
          );

          if (response.isMatch) {
            // response content & schema diff
            const responseContentMatch =
              responsesMatcher.responseContentDiffsForTraffic(
                traffic,
                response.context
              );

            results.diffs.push(...responseContentMatch.diffs);
          } else results.diffs.push(...response.diffs);

          // collect diffs from failed matches
          return results;
        } else return { diffs: matchesOperation.diffs, errors: [] };
      }
      return { diffs: matchesPath.diffs, errors: [] };
    },
  };
}
