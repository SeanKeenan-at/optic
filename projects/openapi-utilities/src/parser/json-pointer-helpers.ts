import * as jsonPointer from "json-pointer";

function append(pointer: string, ...property: string[]): string {
  const parsed = jsonPointer.parse(pointer.toString()) || [];
  return jsonPointer.compile([...parsed, ...property]);
}

function pop(pointer: string): string {
  const parsed = jsonPointer.parse(pointer.toString()) || [];
  parsed.pop();
  return jsonPointer.compile([...parsed]);
}

function unescapeUriSafePointer(inputFromApiToolkit: string): string {
  return decodeURIComponent(inputFromApiToolkit);
}

function compile(input: string[]) {
  return jsonPointer.compile(input);
}

function decode(pointer: string): string[] {
  return jsonPointer.parse(pointer);
}

export default {
  append,
  pop,
  decode,
  unescapeUriSafePointer,
  get: jsonPointer.get,
  compile,
};
