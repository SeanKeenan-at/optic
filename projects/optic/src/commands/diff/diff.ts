import { Command } from 'commander';
import brotli from 'brotli';
import open from 'open';

import {
  generateSpecResults,
  logComparison,
  generateChangelogData,
  terminalChangelog,
} from '@useoptic/openapi-utilities';
import { wrapActionHandlerWithSentry } from '@useoptic/openapi-utilities/build/utilities/sentry';
import {
  BreakingChangesRuleset,
  NamingChangesRuleset,
} from '@useoptic/standard-rulesets';
import { RuleRunner, Ruleset } from '@useoptic/rulesets-base';
import {
  parseFilesFromRef,
  ParseResult,
  getFileFromFsOrGit,
} from '../../utils/spec-loaders';
import { OpticCliConfig, VCS } from '../../config';
import chalk from 'chalk';
import {
  flushEvents,
  trackEvent,
} from '@useoptic/openapi-utilities/build/utilities/segment';
import { getAnonId } from '../../utils/anonymous-id';

const description = `run a diff between two API specs`;

const usage = () => `
  optic diff --id user-api --base <base>
  optic diff <file_path> --base <base>
  optic diff <file_path> <file_to_compare_against>`;

const helpText = `
Example usage:
  Run a diff against the api spec \`user-api\` using the config from your \`optic.yml\` file against master
  $ optic diff --id user-api --base master

  Run a diff between \`master:specs/openapi-spec.yml\` and \`specs/openapi-spec.yml\`
  $ optic diff openapi-spec.yml --base master

  Run a diff between \`openapi-spec-v0.yml\` and \`openapi-spec-v1.yml\`
  $ optic diff openapi-spec-v0.yml openapi-spec-v1.yml
  
  Run a diff and view changes in the Optic web view
  $ optic diff --id user-api --base master --web`;

type SpecResults = Awaited<ReturnType<typeof generateSpecResults>>;
const webBase =
  process.env.OPTIC_ENV === 'staging'
    ? 'https://app.o3c.info'
    : process.env.OPTIC_ENV === 'local'
    ? 'http://localhost:3000'
    : 'https://app.useoptic.com';

const stdRulesets = {
  'breaking-changes': BreakingChangesRuleset,
  // 'naming-changes': NamingChangesRuleset,
};

export const registerDiff = (cli: Command, config: OpticCliConfig) => {
  cli
    .command('diff')
    .configureHelp({
      commandUsage: usage,
    })
    .addHelpText('after', helpText)
    .description(description)
    .argument('[file_path]', 'path to file to compare')
    .argument('[file_to_compare_against]', 'path to file to compare with')
    .option(
      '--base <base>',
      'the base ref to compare against. Defaults to master',
      'master'
    )
    .option(
      '--id <id>',
      'the id of the spec to run against in defined in the `optic.yml` file'
    )
    .option('--no-checks', 'disable checks')
    .option('--web', 'view the diff in the optic changelog web view', false)
    .action(
      wrapActionHandlerWithSentry(
        async (
          file1: string | undefined,
          file2: string | undefined,
          options: {
            base: string;
            id?: string;
            checks: boolean;
            web: boolean;
          }
        ) => {
          let baseFile: ParseResult;
          let headFile: ParseResult;

          if (file1 && file2) {
            const baseFilePath = file1;
            const headFilePath = file2;
            [baseFile, headFile] = await Promise.all([
              getFileFromFsOrGit(baseFilePath),
              getFileFromFsOrGit(headFilePath),
            ]);
          } else if (file1) {
            const commandVariant = `optic diff <file> --base <ref>`;
            if (config.vcs !== VCS.Git) {
              console.error(
                `Error: ${commandVariant} must be called from a git repository.`
              );
              return;
            }

            ({ baseFile, headFile } = await parseFilesFromRef(
              file1,
              options.base,
              config.root
            ));
          } else if (options.id) {
            const commandVariant = `optic diff --id <id> --base <ref>`;
            if (config.vcs !== VCS.Git) {
              console.error(
                `Error: ${commandVariant} must be called from a git repository.`
              );
              return;
            }
            if (!config.configPath) {
              console.error(
                `Error: no optic.yml config file was found. optic.yml must be included for ${commandVariant}`
              );
              return;
            }

            console.log('Running diff against files from optic.yml file');
            const files = config.files;
            const maybeMatchingFile = files.find(
              (file) => file.id === options.id
            );

            if (maybeMatchingFile) {
              ({ baseFile, headFile } = await parseFilesFromRef(
                maybeMatchingFile.path,
                options.base,
                config.root
              ));
            } else {
              console.error(
                `id: ${options.id} was not found in the optic.yml file`
              );
              console.log(
                `valid list of file names: ${files
                  .map((file) => file.id)
                  .join(', ')}`
              );
              return;
            }
          } else {
            console.error('Invalid combination of arguments');
            console.log(helpText);
            return;
          }

          const ruleRunner = generateRuleRunner(config, options.checks);
          const specResults = await generateSpecResults(
            ruleRunner,
            baseFile,
            headFile,
            null
          );

          const changelogData = generateChangelogData({
            changes: specResults.changes,
            toFile: headFile.jsonLike,
          });

          console.log('');
          for (const log of terminalChangelog(changelogData)) {
            console.log(log);
          }

          if (options.checks) {
            if (specResults.results.length > 0) {
              console.log('Checks');
              console.log('');
            }

            logComparison(specResults, { output: 'pretty', verbose: false });
          }

          if (options.web) {
            const compressedData = compressData(
              baseFile,
              headFile,
              specResults
            );
            console.log('Opening up diff in web view');
            const anonymousId = await getAnonId();
            trackEvent('optic.diff.view_web', anonymousId, {
              compressedDataLength: compressedData.length,
            });
            await flushEvents();
            await openBrowserToPage(`${webBase}/cli/diff#${compressedData}`);
          } else {
            console.log(
              chalk.blue(
                `Rerun this command with the --web flag to view these changes to view the detailed changes`
              )
            );
          }
        }
      )
    );
};

const compressData = (
  baseFile: ParseResult,
  headFile: ParseResult,
  specResults: SpecResults
): string => {
  const dataToCompress = {
    base: baseFile.jsonLike,
    head: headFile.jsonLike,
    results: specResults,
    version: '1',
  };
  // TODO maybe strip out unnecessary things here?
  // We could strip out:
  // - components that do not have a `$ref` key - they should be flattened, except for any circular refs
  const compressed = brotli.compress(
    Buffer.from(JSON.stringify(dataToCompress))
  );
  const urlSafeString = Buffer.from(compressed).toString('base64');
  return urlSafeString;
};

const openBrowserToPage = async (url: string) => {
  await open(url, { wait: false });
};

const generateRuleRunner = (
  config: OpticCliConfig,
  checksEnabled: boolean
): RuleRunner => {
  const rulesets: Ruleset[] = [];

  if (checksEnabled) {
    for (const rule of config.ruleset) {
      if (typeof rule === 'string' && stdRulesets[rule]) {
        rulesets.push(new stdRulesets[rule]());
      } else {
        console.error(`Warning: Invalid ruleset ${rule}`);
      }
    }
  }

  return new RuleRunner(rulesets);
};
