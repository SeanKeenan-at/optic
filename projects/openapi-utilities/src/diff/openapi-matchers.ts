import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
import { getParameterIdentity } from './array-identifiers';

export const isPathParameterArray = (pointer: string): boolean => {
  return (
    jsonPointerHelpers.matches(pointer, ['paths', '**', 'parameters']) ||
    jsonPointerHelpers.matches(pointer, [
      'paths',
      '**',
      '!parameters',
      'parameters',
    ])
  );
};

export const isPathsMap = (pointer: string): boolean => {
  return jsonPointerHelpers.matches(pointer, ['paths']);
};
