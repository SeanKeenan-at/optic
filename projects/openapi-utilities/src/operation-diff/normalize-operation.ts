import { OpenAPIV3 } from 'openapi-types';
import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
import { Coordinates } from './coordinates';
import JsonPointerHelpers from '@useoptic/json-pointer-helpers/build/json-pointers/json-pointer-helpers';

export function normalizeOperation(
  spec: OpenAPIV3.Document,
  method: OpenAPIV3.HttpMethods,
  path: string
): OpenAPIOperation {
  const coordinateAliases: {
    [key: string]: string;
  } = {};

  const operationCoordinate = jsonPointerHelpers.compile([
    'paths',
    path,
    method,
  ]);

  let operation = jsonPointerHelpers.get(spec, operationCoordinate);

  // in-source security
  if (spec.security && !operation.security) {
    operation = { ...operation, security: spec.security };
    coordinateAliases[
      JsonPointerHelpers.append(operationCoordinate, 'security')
    ] = JsonPointerHelpers.compile(['security']);
  }

  if (spec.paths[path]!.parameters) {
    const sharedParams = spec.paths[path]!.parameters || [];
    const operationParamsCount = (operation.parameters || []).length;
    operation = {
      ...operation,
      parameters: [...(operation.parameters || []), ...sharedParams],
    };
    sharedParams.forEach((sharedParameter, index) => {
      coordinateAliases[
        JsonPointerHelpers.append(
          operationCoordinate,
          'parameters',
          (operationParamsCount + index).toString()
        )
      ] = jsonPointerHelpers.compile([
        'paths',
        path,
        'parameters',
        index.toString(),
      ]);
    });
  }

  return {
    method,
    path,
    context: {
      environments: spec.servers || [],
      securitySchemas: spec?.components?.securitySchemes as any, // need to fork openapi-type and get rid of $ref
      sharedParameters: spec.paths[path]!.parameters || ([] as any), // need to fork openapi-type and get rid of $ref,
    },
    operation,
    coordinates: {
      absoluteJsonPath: jsonPointerHelpers.compile(['paths', path, method]),
    },
    coordinateAliases,
  };
}

export interface OpenAPIOperation {
  method: string;
  path: string;
  context: {
    environments: OpenAPIV3.ServerObject[];
    securitySchemas: {
      [key: string]: OpenAPIV3.SecuritySchemeObject;
    };
    sharedParameters: OpenAPIV3.ParameterObject[];
  };
  operation: OpenAPIV3.OperationObject;
  coordinates: Coordinates;
  // for post-processing rule results
  coordinateAliases: {
    [key: string]: string;
  };
}
