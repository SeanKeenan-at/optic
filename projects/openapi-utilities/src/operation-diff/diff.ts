import { OpenAPIOperation } from './normalize-operation';
import { OpenApiKind } from '../openapi3/sdk/types';
import { diffCollectionByComputedIndex } from './diff-collection-by-computed-index';
import { OpenAPIV3 } from 'openapi-types';
import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';

export function diff(
  before: OpenAPIOperation,
  after: OpenAPIOperation
): OperationDiffResult[] {
  const differ = new OperationDiff(before, after);
  return differ.diff();
}

class OperationDiff {
  constructor(
    private before: OpenAPIOperation,
    private after: OpenAPIOperation
  ) {}

  private results: OperationDiffResult[] = [];

  diff() {
    this.results = [];
    this.diffParameters();
    return this.results;
  }

  private emitAdded(kind: OpenApiKind, apiId: string, jsonPath: string) {
    this.results.push({
      diff: 'node-added',
      kind,
      coordinates: {
        jsonPath,
        apiId,
      },
    });
  }
  private emitRemoved(kind: OpenApiKind, apiId: string, jsonPath: string) {
    this.results.push({
      diff: 'node-removed',
      kind,
      coordinates: {
        jsonPath,
        apiId,
      },
    });
  }
  private emitUpdated(
    kind: OpenApiKind,
    apiId: string,
    jsonPath: string,
    change: { attribute: string; before: any; after: any }
  ) {
    this.results.push({
      diff: 'node-updated',
      kind,
      coordinates: {
        jsonPath,
        apiId,
      },
      ...change,
    });
  }

  ////

  diffParameters() {
    const parameterDiff =
      diffCollectionByComputedIndex<OpenAPIV3.ParameterObject>(
        (param) => {
          return `${param.in}/${param.name}`;
        },
        (this.before.operation.parameters || []) as OpenAPIV3.ParameterObject[],
        (this.after.operation.parameters || []) as OpenAPIV3.ParameterObject[]
      );

    const inToKind = (inValue: OpenAPIV3.ParameterObject['in']) => {
      if (inValue === 'query') return OpenApiKind.QueryParameter;
      if (inValue === 'path') return OpenApiKind.PathParameter;
      if (inValue === 'header') return OpenApiKind.HeaderParameter;
      if (inValue === 'cookie') return OpenApiKind.CookieParameter;
      throw new Error(`${inValue} parameter invalid`);
    };

    parameterDiff.addedNodes().forEach(([id, node]) => {
      this.emitAdded(
        inToKind(node.in),
        id,
        jsonPointerHelpers.append(
          this.after.coordinates.absoluteJsonPath,
          'parameters',
          (this.after.operation.parameters || []).indexOf(node).toString()
        )
      );
    });

    parameterDiff.removedNodes().forEach(([id, node]) => {
      this.emitRemoved(
        inToKind(node.in),
        id,
        jsonPointerHelpers.append(
          this.before.coordinates.absoluteJsonPath,
          'parameters',
          (this.before.operation.parameters || []).indexOf(node).toString()
        )
      );
    });

    parameterDiff.continuousNodes().forEach(([id, before, after]) => {
      diffAttributes(before, after, ['name', 'in', 'schema']).forEach(
        (nodeUpdate) => {
          this.emitUpdated(
            inToKind(before.in), //
            id,
            jsonPointerHelpers.append(
              this.after.coordinates.absoluteJsonPath,
              'parameters',
              (this.after.operation.parameters || []).indexOf(after).toString(),
              nodeUpdate.attribute
            ),
            nodeUpdate
          );

          // @todo visit the schema
        }
      );
    });
  }

  visitParameter() {}

  diffResponses() {}
  visitResponse() {}

  diffRequests() {}
  visitRequest() {}

  visitSchema() {}
}

function diffAttributes(before: object, after: object, exclude: string[]) {
  const attributeDiff = diffCollectionByComputedIndex<[string, any]>(
    ([key]) => {
      return key;
    },
    Object.entries(before).filter(([key]) => !exclude.includes(key)),
    Object.entries(after).filter(([key]) => !exclude.includes(key))
  );

  const results: { attribute: string; before: any; after: any }[] = [];

  attributeDiff.addedNodes().forEach(([id, addedValue]) => {
    results.push({ attribute: id, before: undefined, after: addedValue[1] });
  });

  attributeDiff.removedNodes().forEach(([id, removedValue]) => {
    results.push({ attribute: id, before: removedValue[1], after: undefined });
  });

  attributeDiff.continuousNodes().forEach(([id, before, after]) => {
    results.push({ attribute: id, before: before[1], after: after[1] });
  });

  return results;
}

type OperationDiffResult = {
  kind: OpenApiKind;
  coordinates: {
    apiId: string;
    jsonPath: string;
  };
} & (
  | {
      diff: 'node-added';
    }
  | {
      diff: 'node-removed';
    }
  | { diff: 'node-updated'; attribute: string; before: any; after: any }
);
