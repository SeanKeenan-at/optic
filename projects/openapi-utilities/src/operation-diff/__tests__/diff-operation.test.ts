import fs from 'fs-extra';
import { normalizeOperation, OpenAPIOperation } from '../normalize-operation';
import { OpenAPIV3 } from 'openapi-types';
import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
import { diff } from '../diff';

const jsonFromFile = async (path: string) => {
  const bytes = await fs.readJson(path);
  return bytes;
};

function opWithParameters(
  parameters: OpenAPIV3.ParameterObject[]
): OpenAPIOperation {
  return {
    method: OpenAPIV3.HttpMethods.GET,
    path: '/example',
    coordinates: {
      absoluteJsonPath: jsonPointerHelpers.compile([
        'paths',
        '/example',
        'get',
      ]),
    },
    context: {
      sharedParameters: [],
      securitySchemas: {},
      environments: [],
    },
    coordinateAliases: {},
    operation: {
      parameters,
      responses: {},
      summary: '',
    },
  };
}

it('can normalize an openapi operation', async () => {
  const beforeOp = opWithParameters([
    { in: 'query', name: 'example', required: true, explode: true },
    {
      in: 'path',
      name: 'id',
      required: true,
      schema: {
        type: 'string',
      },
    },
  ]);
  const afterOp = opWithParameters([
    { in: 'query', name: 'example', required: false, description: 'this one' },
    {
      in: 'query',
      name: 'other-query',
      required: false,
      schema: {
        type: 'string',
      },
    },
  ]);
  const results = diff(beforeOp, afterOp);

  expect(results).toMatchSnapshot();
});
