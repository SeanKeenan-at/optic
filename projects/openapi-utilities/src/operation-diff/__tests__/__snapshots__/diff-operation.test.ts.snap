// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can normalize an openapi operation 1`] = `
Array [
  Object {
    "coordinates": Object {
      "apiId": "query/other-query",
      "jsonPath": "/paths/~1example/get/parameters/1",
    },
    "diff": "node-added",
    "kind": "query-parameter",
  },
  Object {
    "coordinates": Object {
      "apiId": "path/id",
      "jsonPath": "/paths/~1example/get/parameters/1",
    },
    "diff": "node-removed",
    "kind": "path-parameter",
  },
  Object {
    "after": "this one",
    "attribute": "description",
    "before": undefined,
    "coordinates": Object {
      "apiId": "query/example",
      "jsonPath": "/paths/~1example/get/parameters/0/description",
    },
    "diff": "node-updated",
    "kind": "query-parameter",
  },
  Object {
    "after": undefined,
    "attribute": "explode",
    "before": true,
    "coordinates": Object {
      "apiId": "query/example",
      "jsonPath": "/paths/~1example/get/parameters/0/explode",
    },
    "diff": "node-updated",
    "kind": "query-parameter",
  },
  Object {
    "after": false,
    "attribute": "required",
    "before": true,
    "coordinates": Object {
      "apiId": "query/example",
      "jsonPath": "/paths/~1example/get/parameters/0/required",
    },
    "diff": "node-updated",
    "kind": "query-parameter",
  },
]
`;
